src/SNLS_HybrdTrDLDenseG.h

```diff
@@ -0,0 +1,599 @@
+            if (m_nslow2 == 5) {
+                m_status = slowJacobian;
+                return m_status;
+            }
+            if (m_nslow1 == 10) {
+                m_status = slowConvergence;
+                return m_status;
+            }
+            // Only calculate this if solution wasn't rejected
+            if( !reject_prev ) 
+            {
+                // Here we can use delx, nrStep, and grad as working arrays as we're just going
+                // to rewrite them in a second...
+                // nrStep = (R * delx + Q^T * f_i)
+                snls::linalg::matUTriVecMult<m_nDim, m_nDim>(Jacobian, delx, nrStep);
+                for (int i = 0; i < m_nDim; i++) {
+                    // work_arr3 = R \Delta x + Q^T F
+                    // where F here is our residual vector
+                    nrStep[i] += qtf[i];
+                }
+                // calculate the rank one modification to the jacobian
+                // and update qtf if necessary
+                {
+                    // delx = delx / ||delx||_L2
+                    const double idxnorm = 1.0 / snls::linalg::norm<m_nDim>(delx);
+                    for (int i = 0; i < m_nDim; i++) {
+                        delx[i] = delx[i] * idxnorm;
+                    }
+                    snls::linalg::matTVecMult<m_nDim, m_nDim>(Q, residual, grad);
+                    for (int i = 0; i < m_nDim; i++) {
+                        qtf[i] = grad[i];
+                    }
+                    // grad = (Q^T * f_{i+1} - Q^T * f_i - R * delx)
+                    for (int i = 0; i < m_nDim; i++) {
+                        grad[i] = (grad[i] - nrStep[i]) * idxnorm;
+                    }
+                }
+
+                // compute the qr factorization of the updated jacobian
+                bool singular = false;
+                this->rank1_update(delx, grad, Jacobian, Q, qtf, nrStep, singular);
+                if (singular) {
+                    m_status = algFailure;
+                    return m_status;
+                }
+            // 
+            }
+            res_0 = m_res;
+            jeval = false;
+        } // End of while loop
+        // Return the unconverged result
+        return m_status;
+    }
+
+    // This just does the solve of
+    // R x = Q^T f
+    // where we're solving for x
+    // Since R is upper diagonal this  is a trivial process.
+    __snls_hdev__
+    inline
+    bool computeNewtonStep(const double* const rmat,
+                           const double* const func,
+                           double* const x,
+                           const double tol = 1e-50)
+    {
+        const double ndim1 = m_nDim - 1;
+        for (int i = ndim1; i > -1; i-- ) {
+            double temp = rmat[SNLS_NN_INDX(i, i, m_nDim)];
+            // If R has a tiny diagonal diagonal 
+            if (fabs(temp) < tol) {
+                // We could try something like the above to make sure this function always
```

Comment:
You mean "like the coding below" instead of "like the above"?

Reply:
Oops yep I meant "like the coding below"...